src/core/lib/transport/batch_builder.cc

```diff
@@ -0,0 +1,181 @@
+// Copyright 2023 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <grpc/support/port_platform.h>
+
+#include "src/core/lib/transport/batch_builder.h"
+
+#include <string.h>
+
+#include <type_traits>
+
+#include "src/core/lib/promise/poll.h"
+#include "src/core/lib/slice/slice.h"
+#include "src/core/lib/surface/call_trace.h"
+#include "src/core/lib/transport/metadata_batch.h"
+#include "src/core/lib/transport/transport.h"
+#include "src/core/lib/transport/transport_impl.h"
+
+namespace grpc_core {
+
+BatchBuilder::BatchBuilder(grpc_transport_stream_op_batch_payload* payload)
+    : payload_(payload) {}
+
+void BatchBuilder::PendingCompletion::CompletionCallback(
+    void* self, grpc_error_handle error) {
+  auto* pc = static_cast<PendingCompletion*>(self);
+  auto* party = pc->batch->party.get();
+  if (grpc_call_trace.enabled()) {
+    gpr_log(
+        GPR_DEBUG, "%s[connected] Finish batch-component %s for %s: status=%s",
+        party->DebugTag().c_str(), std::string(pc->name()).c_str(),
+        grpc_transport_stream_op_batch_string(&pc->batch->batch, false).c_str(),
+        error.ToString().c_str());
+  }
+  party->Spawn(
+      "batch-completion",
+      [pc, error = std::move(error)]() mutable {
+        RefCountedPtr<Batch> batch = std::exchange(pc->batch, nullptr);
+        pc->done_latch.Set(std::move(error));
+        return Empty{};
+      },
+      [](Empty) {});
+}
+
+BatchBuilder::PendingCompletion::PendingCompletion(RefCountedPtr<Batch> batch)
+    : batch(std::move(batch)) {
+  GRPC_CLOSURE_INIT(&on_done_closure, CompletionCallback, this, nullptr);
+}
+
+BatchBuilder::Batch::Batch(grpc_transport_stream_op_batch_payload* payload,
+                           grpc_stream_refcount* stream_refcount)
+    : party(static_cast<Party*>(Activity::current())->Ref()),
+      stream_refcount(stream_refcount) {
+  memset(&batch, 0, sizeof(batch));
```

Comment:
Don't think this is needed, since `grpc_transport_stream_op_batch` has a C++-style ctor.

Reply:
done